# Jinhua-scRNAseq

## Overview

Understanding how intestinal development shapes whole-body metabolism is essential for explaining age-dependent susceptibility to obesity. Here, we generate the most comprehensive dynamic single-cell atlas of the Jinhua pig intestine (n=25, k=168,759), capturing spatial and temporal heterogeneity, and resolve "Cell-Tissue-Individual" dynamic metabolic states from enterocytes with phenotypic and multi-omics datasets. Functional validation demonstrates that PCK1 acts as a spatiotemporal metabolic regulator, coordinating glycolysis/gluconeogenesis and intersecting with the TCA and pentose phosphate pathways. Cross-tissue analysis provides an insight into the PCK1-centered metabolic multi-organ axis, potentially linked to cholesterol transport and systemic metabolism. Cross-species analysis further highlights an unrecognized influence of PCK1 on obesity susceptibility in both pigs and humans. Furthermore, comparative analysis suggests a close recapitulation in small intestinal biology between neonatal pigs and humans, offering a robust model for metabolic studies. Overall, these results provide a dynamic cellular framework for understanding intestinal development and metabolism, uncover a conserved PCK1-centered metabolic axis underlying obesity risk, and offer a powerful pig-based platform for studying metabolic disorders.

## Workflow

The analyses contained in our work:

- JPISCA: Integrated single-cell atlas of JH intestine across developmental stages.
- Regional and temporal regulation of enterocytes and colonocytes in metabolic characteristics
- Spatiotemporal genes linking carbohydrate metabolism and obesity traits
- PCK1 affects G/G, TCA, and PPP in aged pig intestines and vitro enterocytesã€‚
- Intestinal PCK1 gene engages multicellular ecosystem and obesity regulation: multi-omics evidence of genetic variation and causal relationship.
- Application of cross-species intestinal models, taking enterocytes and colonocytes as examples.

## Data source

All single-cell datasets were available for interactive exploration and download in our interactive web portal (http://alphaindex.zju.edu.cn/PigCellDB). Genomic datasets were collected from PHARP (http://alphaindex.zju.edu.cn/PHARP). All other data of different formats were available from the corresponding author upon reasonable request. Source data are provided with this paper. The code generated during this study will be available at GitHub https://github.com/Qinqin-Xie/Jinhua-scRNAseq.

## License

Panlab-QinqinXie

[//]: # (These are reference links used in the body of this note and get stripped out when the markdown processor does its job. There is no need to format nicely because it shouldn't be seen. Thanks SO - http://stackoverflow.com/questions/4823468/store-comments-in-markdown-syntax)

   [dill]: <https://github.com/joemccann/dillinger>

   [git-repo-url]: <https://github.com/joemccann/dillinger.git>

   [john gruber]: <http://daringfireball.net>

   [df1]: <http://daringfireball.net/projects/markdown/>

   [markdown-it]: <https://github.com/markdown-it/markdown-it>

   [Ace Editor]: <http://ace.ajax.org>

   [node.js]: <http://nodejs.org>

   [Twitter Bootstrap]: <http://twitter.github.com/bootstrap/>

   [jQuery]: <http://jquery.com>

   [@tjholowaychuk]: <http://twitter.com/tjholowaychuk>

   [express]: <http://expressjs.com>

   [AngularJS]: <http://angularjs.org>

   [Gulp]: <http://gulpjs.com>

   [PlDb]: <https://github.com/joemccann/dillinger/tree/master/plugins/dropbox/README.md>

   [PlGh]: <https://github.com/joemccann/dillinger/tree/master/plugins/github/README.md>

   [PlGd]: <https://github.com/joemccann/dillinger/tree/master/plugins/googledrive/README.md>

   [PlOd]: <https://github.com/joemccann/dillinger/tree/master/plugins/onedrive/README.md>

   [PlMe]: <https://github.com/joemccann/dillinger/tree/master/plugins/medium/README.md>

   [PlGa]: <https://github.com/RahulHP/dillinger/blob/master/plugins/googleanalytics/README.md>
